akka-stream/src/main/scala/akka/stream/scaladsl/Tcp.scala

```diff
@@ -208,6 +219,89 @@ final class Tcp(system: ExtendedActorSystem) extends akka.actor.Extension {
    */
   def outgoingConnection(host: String, port: Int): Flow[ByteString, ByteString, Future[OutgoingConnection]] =
     outgoingConnection(InetSocketAddress.createUnresolved(host, port))
+
+  /**
+   * Creates an [[Tcp.OutgoingConnection]] with TLS.
+   * The returned flow represents a TCP client connection to the given endpoint where all bytes in and
+   * out go through TLS.
+   *
+   * For more advanced use cases you can manually combine [[Tcp.outgoingConnection()]] and [[TLS]]
+   *
+   * @see [[Tcp.outgoingConnection()]]
+   */
+  def outgoingTlsConnection(
+    host:                String,
+    port:                Int,
+    sslContext:          SSLContext,
+    negotiateNewSession: NegotiateNewSession): Flow[ByteString, ByteString, Future[OutgoingConnection]] =
+    outgoingTlsConnection(InetSocketAddress.createUnresolved(host, port), sslContext, negotiateNewSession)
+
+  /**
+   * Creates an [[Tcp.OutgoingConnection]] with TLS.
+   * The returned flow represents a TCP client connection to the given endpoint where all bytes in and
+   * out go through TLS.
+   *
+   * @see [[Tcp.outgoingConnection()]]
+   */
+  def outgoingTlsConnection(
+    remoteAddress:       InetSocketAddress,
+    sslContext:          SSLContext,
+    negotiateNewSession: NegotiateNewSession,
+    localAddress:        Option[InetSocketAddress]           = None,
+    options:             immutable.Traversable[SocketOption] = Nil,
+    connectTimeout:      Duration                            = Duration.Inf,
+    idleTimeout:         Duration                            = Duration.Inf): Flow[ByteString, ByteString, Future[OutgoingConnection]] = {
+
+    val connection = outgoingConnection(remoteAddress, localAddress, options, true, connectTimeout, idleTimeout)
+    val tls = TLS(sslContext, negotiateNewSession, TLSRole.client)
+    connection.join(tlsSupport.atop(tls).reversed)
+  }
+
+  /**
+   * Creates a [[Tcp.ServerBinding]] instance which represents a prospective TCP server binding on the given `endpoint`
+   * where all incoming and outgoing bytes are passed through TLS.
+   *
+   * @see [[Tcp.bind()]]
+   */
+  def bindTls(
+    interface:           String,
+    port:                Int,
+    sslContext:          SSLContext,
+    negotiateNewSession: NegotiateNewSession,
+    backlog:             Int                                 = 100,
+    options:             immutable.Traversable[SocketOption] = Nil,
+    idleTimeout:         Duration                            = Duration.Inf): Source[IncomingConnection, Future[ServerBinding]] = {
+
+    val tls = TLS(sslContext, negotiateNewSession, TLSRole.server)
+
+    bind(interface, port, backlog, options, true, idleTimeout).map { incomingConnection ⇒
+      incomingConnection.copy(
+        flow = incomingConnection.flow.join(tlsSupport.atop(tls).reversed)
```

Comment:
You could move that `tlsSupport.atop(tls).reversed` out of the stream as well.

Reply:
If you want to leave it like this, it could make sense to formulate it a bit differently so that it its clear in which direction things are stacked on top of each other.

Something like

```scala
incomingConnection.flow
   .atop(tls) // will only work in one direction (might need reversed, though)
   .join(tlsSupport) // will only work in one direction (might need reversed, though)
```

Written in that way you cannot accidentally connect the TLS part the wrong way round.